plugins/lang-plan-a/src/main/java/org/elasticsearch/plan/a/Metadata.java

```diff
@@ -0,0 +1,598 @@
+         */
+        Object postConst = null;
+
+        /**
+         * The isNull variable is set to true when a null constant statement is made in the script.  This allows the
+         * {@link Writer} to potentially shortcut certain comparison operations.
+         */
+        boolean isNull = false;
+
+        /**
+         * The to variable is used to track what an ANTLR node's value should be cast to.  This is set on every ANTLR
+         * node in the tree, and used by the {@link Writer} to make a run-time cast if necessary in the byte code.
+         * This is also used by the {@link Analyzer} to determine if a cast is legal.
+         */
+        Type to = null;
+
+        /**
+         * The from variable is used to track what an ANTLR node's value should be cast from.  This is set on every
+         * ANTLR node in the tree independent of other nodes.  This is used by the {@link Analyzer} to determine if a
+         * cast is legal.
+         */
+        Type from = null;
+
+        /**
+         * The explicit variable is set to true when a cast is explicitly made in the script.  This tracks whether
+         * or not a cast is a legal up cast.
+         */
+        boolean explicit = false;
+
+        /**
+         * The typesafe variable is set to true when a dynamic type is used as part of an expression.  This propagates
+         * up the tree to the top of the expression.  This allows for implicit up casts throughout the expression and
+         * is used by the {@link Analyzer}.
+         */
+        boolean typesafe = true;
+
+        /**
+         * This is set to the combination of the to and from variables at the end of each node visit in the
+         * {@link Analyzer}.  This is set on every ANTLR node in the tree independent of other nodes, and is
+         * used by {@link Writer} to make a run-time cast if necessary in the byte code.
+         */
+        Cast cast = null;
+
+        // The constructor takes in the associated ANTLR node.
+        private ExpressionMetadata(final ParserRuleContext source) {
+            this.source = source;
+        }
+    }
+
+    /**
+     * ExternalMetadata is used to store metadata about the overall state of a variable/method chain such as
+     * '(int)x.get(3)' where each piece of that chain is broken into it's indiviual pieces and stored in
+     * {@link ExtNodeMetadata}.
+     */
+    static class ExternalMetadata {
+        /**
+         * The source variable is the ANTLR node used to generate this metadata.
+         */
+        final ParserRuleContext source;
+
+        /**
+         * The read variable is set to true when the value of a variable/method chain is going to be read from.
+         * This is used by the {@link Analyzer} to determine if this variable/method chain will be in a standalone
+         * statement.
+         */
+        boolean read = false;
```

Comment:
You might want to have a superclass here just so common stuff is obvious like `source` and (sometimes) `read`.